server/src/test/java/io/druid/server/coordinator/CuratorDruidCoordinatorTest.java

```diff
@@ -0,0 +1,522 @@
+        new Duration(COORDINATOR_PERIOD),
+        null,
+        null,
+        new Duration(COORDINATOR_PERIOD),
+        null,
+        10,
+        null,
+        false,
+        false,
+        new Duration("PT0s")
+    );
+    sourceLoadQueueChildrenCache = new PathChildrenCache(
+        curator,
+        LOADPATH,
+        true,
+        true,
+        Execs.singleThreaded("coordinator_test_path_children_cache-%d")
+    );
+    destinationLoadQueueChildrenCache = new PathChildrenCache(
+        curator,
+        LOADPATH2,
+        true,
+        true,
+        Execs.singleThreaded("coordinator_test_path_children_cache2-%d")
+    );
+    loadQueuePeon = new CuratorLoadQueuePeon(
+        curator,
+        LOADPATH,
+        objectMapper,
+        Execs.scheduledSingleThreaded("coordinator_test_load_queue_peon_scheduled-%d"),
+        Execs.singleThreaded("coordinator_test_load_queue_peon-%d"),
+        druidCoordinatorConfig
+    );
+    loadQueuePeon2 = new CuratorLoadQueuePeon(
+        curator,
+        LOADPATH2,
+        objectMapper,
+        Execs.scheduledSingleThreaded("coordinator_test_load_queue_peon2_scheduled-%d"),
+        Execs.singleThreaded("coordinator_test_load_queue_peon2-%d"),
+        druidCoordinatorConfig
+    );
+    druidNode = new DruidNode("hey", "what", 1234, null, true, false);
+    loadManagementPeons = new ConcurrentHashMap<>();
+    scheduledExecutorFactory = (corePoolSize, nameFormat) -> Executors.newSingleThreadScheduledExecutor();
+    leaderAnnouncerLatch = new CountDownLatch(1);
+    leaderUnannouncerLatch = new CountDownLatch(1);
+    coordinator = new DruidCoordinator(
+        druidCoordinatorConfig,
+        new ZkPathsConfig()
+        {
+
+          @Override
+          public String getBase()
+          {
+            return "druid";
+          }
+        },
+        configManager,
+        databaseSegmentManager,
+        baseView,
+        metadataRuleManager,
+        curator,
+        new NoopServiceEmitter(),
+        scheduledExecutorFactory,
+        null,
+        null,
+        new NoopServiceAnnouncer()
+        {
+          @Override
+          public void announce(DruidNode node)
+          {
+            // count down when this coordinator becomes the leader
+            leaderAnnouncerLatch.countDown();
+          }
+
+          @Override
+          public void unannounce(DruidNode node)
+          {
+            leaderUnannouncerLatch.countDown();
+          }
+        },
+        druidNode,
+        loadManagementPeons,
+        null,
+        new CostBalancerStrategyFactory(),
+        EasyMock.createNiceMock(LookupCoordinatorManager.class),
+        new TestDruidLeaderSelector()
+    );
+  }
+
+  @After
+  public void tearDown() throws Exception
+  {
+    baseView.stop();
+    loadQueuePeon.stop();
+    sourceLoadQueueChildrenCache.close();
+    destinationLoadQueueChildrenCache.close();
+    tearDownServerAndCurator();
+  }
+
+  @Test
```

Comment:
Please add a timeout to avoid too long waits even when this test becomes broken in the future.